### ISSUE TITLE
Oddity with ":@generic" and code completion

### ISSUE THREAD

**Issue by NotHyper-474** [vshaxe/vshaxe#627]
Reproduction steps:
- Clone the Funkin repo `git clone https://github.com/FunkinCrew/Funkin.git`
- Follow the compile guide in `docs/COMPILING.md` to setup the libraries
- Open `.vscode/settings.json` to change the `"haxe.displayServer` setting to change log print -> completion to true
- Open any file and trigger an `server/invalidate` by modifying a file
- Press Ctrl+Space to trigger auto-complete, it will fail and only show basic automatic snippets.
- Notice that hover hints are also broken and the Haxe server log has a `Recursive @:generic function` error.

The error happens in `FlxSpriteGroup`, a script of one library the project uses.
```Completion Response =
{
  "jsonrpc": "2.0",
  "id": 15,
  "error": {
    "code": 0,
    "message": "Compiler error",
    "data": [
      {
        "severity": 1,
        "location": {
          "file": "D:\\Users\\***\\Downloads\\Funkin-develop\\.haxelib\\flixel\\git\\flixel\\group\\FlxSpriteGroup.hx",
          "range": {
            "start": { "line": 1080, "character": 2 },
            "end": { "line": 1080, "character": 44 }
          }
        },
        "message": "Recursive @:generic function"
      }
    ]
  }
}
Stats = 1 files, 1185 classes, 40 methods, 366 macros
Time spent : 4.703s
```

Which leads to these lines.
```hx
inline function offsetCallback(Offset:FlxPoint)
		transformChildren(offsetTransform, Offset);
```
Removing "@:generic" from `transformChildren` makes it work, although very slowly (which - according to the troubleshooting guide - is caused by too many cached files being invalidated). And the program compiles fine regardless of if it's present or not.
If that part is not removed eventually it'll start throwing a "Stack overflow" message that only gets solved by restarting the language server.
The project in question uses a modified HaxeFlixel, so perhaps there's something to it.

**Comment by kLabz** [comment]
With what version of Haxe compiler?

**Comment by NotHyper-474** [comment]
4.3.4

Here's the full log if it's of any use.
[log.txt](https://github.com/user-attachments/files/15788894/log.txt)

**Comment by NotHyper-474** [comment]
Upgrading to Haxe 4.3.5 seems to have fixed the issue. However I'll tinker more with the project to confirm that did it.

**Comment by kLabz** [comment]
You might want to try nightlies instead. Haxe compilation server has some issues on 4.x, most were fixed on nightlies

**Comment by NotHyper-474** [comment]
> You might want to try nightlies instead. Haxe compilation server has some issues on 4.x, most were fixed on nightlies

Issue persists on `haxe_2024-07-18_4.3_bugfix_bd79571.zip` and as for `haxe_latest.zip` (5.x) it's throwing an error
```
Failed - try fixing the error(s) and restarting the language server:
 ERROR  (unknown position)
    Invalid_argument("List.iter2")
```

**Comment by kLabz** [comment]
Ok, now that's interesting. 
Is that with CodenameEngine?

**Comment by NotHyper-474** [comment]
Nope it's Friday Night Funkin'. I would try CodenameEngine to see if something similar happens but I'm kind of out of disk space

Either way I noticed after the upgrade that it will still throw a "Recursive @:generic function" error but doesn't seem to be from the same file from before.

I don't know which file is throwing it though as it's not showing where anymore.

I find odd how code completion will only stop working after some time and it happens quicker on large files.

**Comment by NotHyper-474** [comment]
Ok I decided to try Codename which recommends Haxe 4.2.5 and code completion seems to work as intended without any other issues. However apparently it's based on Funkin source code from version 0.2.7 which is pretty different from the current version's codebase.

Using Haxe 4.3 on CodeName still shows `Tainted <thing>` (mostly `check_display_file`) but code completion works as well

**Comment by NotHyper-474** [comment]
Managed to replicate this issue on an [OnWorks VM](https://onworks.net) so I think it's safe to say it's not a problem with my machine. Weird how no one seems to have complained about it on the Funkin' repository yet.

I wish I could at least debug this but I have not managed to locally install vshaxe from source.

**Comment by AlexHaxe** [comment]
they have complained on discord a couple weeks back. I pointed them to `@:generic` of `transformChildren` function as the culprit of no completion (didn't realise there was an issue open at the time). 
since they use a custom HaxeFlixel fork I'm not sure what they did with that info (looking at @EliteMasterEric).

I have a HaxeFlixel project that doesn't have that issue, it's running original HaxeFlixel and my project might not be using the parts that would cause the issue.

so maybe try a simple demo project with original HaxeFlixel to see if it works and then switch to modified HaxeFlixel and compare. if both work, then it could be one of the other 20+ libraries FNF uses or it could be in the features that FNF uses from HaxeFlixel lib which a simple demo project isn't using.
if you can manage to cut down the number of sources required to reproduce, then it's getting easier to debug.

**Comment by NotHyper-474** [comment]
Alright, thanks for confirming it's not only me! Coming back to this, I still haven't found the culprit. I tried replicating the issue like you said but didn't managed to.

However, thanks to some changes to vshaxe I finally got a backtrace of the `Stack overflow` part of the problem, and it's something like below (beware, it's REALLY large)
[Backtrace.txt](https://github.com/user-attachments/files/17532109/Backtrace.txt)


It's all coming from Polymod's `HScriptedClassMacro` and classes that use its `@:hscriptClass` metadata. So perhaps there's something wrong going on with this macro that might also be causing the issue with `FlxSpriteGroup`, but I can't really confirm without enough testing.

**Comment by AlexHaxe** [comment]
I don't know why your backtrace is in json which makes it a bit annoying to handle / read.

anyway from what I can see all of your `@:hscriptClass` annotated classes are included multiple times in your stacktrace. it seems like the macro goes through all of your classes in order of appearance and once it's done it starts fresh at the beginning and does it all over again. it's basically an endless recursive loop. 

you need to make your macro **not** do that.

I don't know exactly what happens, since it's far to big of a code base with too many dependencies and unfamiliar mechanisms. you need to investigate why your macro is running without proper end condition or why it doesn't stop after going over your classes once. it's also weird that it doesn't iterate through your classes in a normal way, but rather in a recursive fashion - a loop that doesn't end is bad, because it freezes your software, a recursive loop that doesn't end crashes your software.

in conclusion currently it looks like a coding error not something vshaxe or Haxe compiler can do anything about.
I might be wrong, since there is a chance the recursion could potentially be something Haxe compiler does.

**Comment by NotHyper-474** [comment]
Oh about that, it's just how it's printed in the log, I also had run it through a prettifier since it looked less cluttered to me than everything being in a single line.

Either way, it can't really be called _my_ macro since it comes from Polymod, but I guess I can just make a separate issue on their repository (actually https://github.com/larsiusprime/polymod/issues/157 seems to be the same to what's happening here, except it's when compiling), and it seems like it isn't the first time Polymod has some weird Stack overflow macro issue (refer to https://github.com/FunkinCrew/funkinBounties/issues/4) and interestingly enough this one also happens with a `@:hscriptClass` annotated class. Therefore, if even the current maintainer of Polymod couldn't figure it out I don't see who can.

So yeah you might be right that it's not an issue Haxe can do something about, it's not a big deal anyway since for the first error I can just comment out a line which is just a bit inconvenient.

**Comment by AlexHaxe** [comment]
> Either way, it can't really be called _my_ macro since it comes from Polymod, but I guess I can just make a separate issue on their repository (actually [larsiusprime/polymod#157](https://github.com/larsiusprime/polymod/issues/157) seems to be the same to what's happening here, except it's when compiling), and it seems like it isn't the first time Polymod has some weird Stack overflow macro issue (refer to [FunkinCrew/funkinBounties#4](https://github.com/FunkinCrew/funkinBounties/issues/4)) and interestingly enough this one also happens with a `@:hscriptClass` annotated class. Therefore, if even the current maintainer of Polymod couldn't figure it out I don't see who can.

I have a strong feeling all three are the same issue - just needs someone to reduce the code to a minimal repro.

I called it your macro since you are the one using it, and as far as I can see the original maintainer has not spoken on it - so you bascially inherited it :).

there is a chance that the error is only reproducible when using compilation server since that seems to be a common denominator.

**Comment by NotHyper-474** [comment]
> I have a strong feeling all three are the same issue - just needs someone to reduce the code to a minimal repro.

You know what, you might be right. I also managed to reproduce the `Stack overflow` problem when compiling with Lime by using the `--connect 6000` flag like the issue I linked from Polymod did, so what I think is happening in these cases is:
1. They'd change some code around and trigger the `Recursive @:generic function` error but not notice it except for no code completion occurring (and actually no hover hints as well, mind you), then modify code again enough times to trigger the `Stack overflow` error, but still not notice it.
2. They would try to build their project with the flag and it would fail since the server is "corrupted".

Funkin' doesn't appear to use that flag anymore (albeit I don't know if it even used to) which is probably why Eric didn't run into that bug anymore, but demo Flixel projects do use it which makes me believe Funkin' did too.
Interestingly just trying to get hover hints enough times triggers the Stack overflow bug:

https://github.com/user-attachments/assets/50e58701-dc35-4c97-9892-61f66d660f39

Either way I'm trying these Polymod scriptable classes in a demo project, can't replicate it for now but maybe if I try copying enough classes from Funkin' I might be able to replicate on it too.

> there is a chance that the error is only reproducible when using compilation server since that seems to be a common denominator.

I don't really know what goes on in the server, are you referring to the `haxe.useCompletionServer` setting? It being on or off doesn't seem to make a difference. Only thing I did notice is that the recursive generic func bug always happens after a `server/invalidate`.

**Comment by NotHyper-474** [comment]
After messing around with the macro I'm pretty certain both issues come from it. Albeit I have a strong feeling that the compilation server might have something to do with it as well, since it only seems to happen when modifying a file which seems to trigger a `server/invalidate`

Anyways I think this should further be tracked in either Polymod or Haxe so I'll be closing this.

**Comment by EliteMasterEric** [comment]
@NotHyper-474 Sorry I never replied to this LOL. For the record, I am the author of the macro in question.

The issue likely exists in part of the code which parses each type implementing `HScriptable`, gets a list of all its method fields recursively, and overrides them to query the attached script. This requires a bunch of type parsing and the error may be specifically due to one of the types of the parent classes of one of the HScriptables.

Not sure if we can narrow it down more without a reliable reproduction though.
